src/main/java/org/testng/DependencyMap.java

```diff
@@ -21,7 +22,20 @@
 
   public DependencyMap(ITestNGMethod[] methods) {
     for (ITestNGMethod m : methods) {
-    	m_dependencies.put( m.getRealClass().getName() + "." +  m.getMethodName(), m);
+      m_dependencies.put( m.getRealClass().getName() + "." +  m.getMethodName(), m);
```

Comment:
I will try to have a deeper look into your PR (which looks nicer) asap. But already, thank for your work.

---------

About your question, I think the answer will be complicate and maybe @cbeust can help us to make a choice.

In my opinion, we can suppose the user want to override the behavior of `ClassA#methodB`, then `ClassA#methodA` will depend on `ClassB#methodB` .
But what about `ClassA#methodB`? Should we ignore it? run it before or after?

We can suppose that it's just a user error who didn't see the first `methodB` too.

Anyway, we can say that overrinding test method is a bad practice and some warnings should be displayed (or even forbit it).